File: hive-mind/transport.py
import asyncio
from abc import ABC, abstractmethod


class Transport(ABC):
    def __init__(self):
        self.on_receive_callback = None

    def init(self):
        pass

    @abstractmethod
    async def send(self, envelope):
        pass

    def on_receive(self, callback):
        if type(callback) is not 'function':
            raise TypeError(f'callback must be a function, instead of {type(callback)}')

        self.on_receive_callback = callback

    async def shutdown(self):
        pass

File: hive-mind/transports/transport.js
class Transport {
	constructor() {
		this._onReceiveCallback = null;
	}

	async init() {
	}

	/**
	 * @param {Envelope} envelope
	 * @abstract 
	 */
	async send(envelope) {
		throw new Error('Not implemented');
	}

	/**
	 * @param {Function} callback
	 * @throws {Error} 
	 */
	onRecieve(callback) { // TODO typo => onReceive
		if (typeof callback !== 'function') {
			throw new Error('callback must be a function');
		}

		this._onReceiveCallback = callback;
	}

	async shutdown() {
	}
	
}

module.exports = Transport;
